controller-client/src/paginator.rs

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+use crate::{ControllerClient, ControllerError, ResultRetry};
+use futures::prelude::*;
+use futures::stream::{self};
+use pravega_client_retry::retry_result::RetryError;
+use pravega_client_shared::Scope;
+use std::vec::IntoIter;
+use tracing::info;
+
+///
+///Helper method to iterated over the all the Pravega streams under the provided Scope.
+///This method returns a stream of values,Pravega streams, produced asynchronously.
+///
+/// The below snippets show case the example uses.
+/// Sample 1:
+///```ignore
+/// # futures::executor::block_on(async {
+/// use pravega_client_shared::Scope;
+/// use pravega_controller_client::paginator::list_streams;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client_config::MOCK_CONTROLLER_URI;
+///     let config = ClientConfigBuilder::default()
+///         .controller_uri(MOCK_CONTROLLER_URI)
+///         .build()
+///         .expect("creating config");
+///     let controller_client = ClientFactory::new(config).get_controller_client();
+///     let stream = list_streams(
+///         Scope {
+///             name: "testScope".to_string(),
+///         },
+///         controller_client,
+///     );
+///     // collect all the Streams in a single vector
+///     let stream_list:Vec<String> = stream.map(|str| str.unwrap()).collect::<Vec<String>>().await;
+///  # });
+/// ```
+///
+/// Sample 2:
+/// ```ignore
+/// # futures::executor::block_on(async {
+/// use pravega_client_shared::Scope;
+/// use pravega_controller_client::paginator::list_streams;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client_config::MOCK_CONTROLLER_URI;
+/// use futures::StreamExt;
+///     let config = ClientConfigBuilder::default()
+///         .controller_uri(MOCK_CONTROLLER_URI)
+///         .build()
+///         .expect("creating config");
+///     let controller_client = ClientFactory::new(config).get_controller_client();
+///     let mut stream = list_streams(
+///         Scope {
+///             name: "testScope".to_string(),
+///         },
+///         controller_client,
+///     );
+/// let pravega_stream_1 = stream.next().await;
+/// let pravega_stream_2 = stream.next().await;
+/// // A None is returned at the end of the stream.
+///  # });
+/// ```
+///
+pub fn list_streams(
+    scope: Scope,
+    client: &dyn ControllerClient,
+) -> impl Stream<Item = Result<String, RetryError<ControllerError>>> + '_ {
```

Comment:
Should this be a `Stream` rather than a string?

Reply:
changed it to ScopedStream